### PR TITLE
add npair loss op

### PR DESCRIPTION
**Pull request by ceci3**
http://www.nec-labs.com/uploads/images/Department-Images/MediaAnalytics/papers/nips16_npairmetriclearning.pdf

### CODE REVIEW

**Review by wanghaoshuang — DISMISSED** [review]
LGTM.

**Comment by lfchener** [comment]
image

File: python/paddle/fluid/layers/nn.py [hunk 1]
```diff
@@ -10560,3 +10561,52 @@ def tree_conv(nodes_vector,
     else:
         pre_activation = out
     return helper.append_activation(pre_activation)
+
+
+def npair_loss(anchor, positive, labels, l2_reg=0.002):
+    '''
+  **Npair Loss Layer**
+  
+  see http://www.nec-labs.com/uploads/images/Department-Images/MediaAnalytics/papers/nips16_npairmetriclearning.pdf
```
Comment (haowang101779990) [inline python/paddle/fluid/layers/nn.py]: To make it more elegant, use indirect link in the manner of:
```suggestion
  Read `Improved Deep Metric Learning with Multi class N pair Loss Objective <http://www.nec-labs.com/uploads/images/Department-Images/MediaAnalytics/papers/nips16_npairmetriclearning.pdf>`_ .
```
Could you please preview the effect again to make sure the link is correct and elegant. The '-' may result in failure, remove them.

File: python/paddle/fluid/layers/nn.py [hunk 2]
```diff
@@ -10560,3 +10561,52 @@ def tree_conv(nodes_vector,
...
+
+  Npair loss requires paired data. Npair loss has two parts, the first part is L2 
+  regularizer on the embedding vector, the second part is cross entropy loss which
+  takes the similarity matrix of anchor and positive as logits.
+
+  Args:
+    anchor(Variable): embedding vector for the anchor image. shape=[batch_size, embedding_dims]
+    positive(Variable): embedding vector for the positive image. shape=[batch_size, embedding_dims]
+    labels(Varieble): 1-D tensor. shape=[batch_size]
+    l2_res(float32): L2 regularization term on embedding vector, default: 0.02
```
Comment (haowang101779990) [inline python/paddle/fluid/layers/nn.py]:
```suggestion
    l2_reg(float32): L2 regularization term on embedding vector, default: 0.02
```

File: python/paddle/fluid/layers/nn.py [hunk 3]
```diff
@@ -10560,3 +10561,52 @@ def tree_conv(nodes_vector,
...
+  see http://www.nec-labs.com/uploads/images/Department-Images/MediaAnalytics/papers/nips16_npairmetriclearning.pdf
+
+  Npair loss requires paired data. Npair loss has two parts, the first part is L2 
+  regularizer on the embedding vector, the second part is cross entropy loss which
+  takes the similarity matrix of anchor and positive as logits.
+
+  Args:
+    anchor(Variable): embedding vector for the anchor image. shape=[batch_size, embedding_dims]
+    positive(Variable): embedding vector for the positive image. shape=[batch_size, embedding_dims]
+    labels(Varieble): 1-D tensor. shape=[batch_size]
```
Comment (haowang101779990) [inline python/paddle/fluid/layers/nn.py]:
```suggestion
    labels(Variable): 1-D tensor. shape=[batch_size]
```

File: python/paddle/fluid/layers/nn.py [hunk 4]
```diff
@@ -10560,3 +10561,52 @@ def tree_conv(nodes_vector,
...
     return helper.append_activation(pre_activation)
+
+
+def npair_loss(anchor, positive, labels, l2_reg=0.002):
+    '''
+  **Npair Loss Layer**
+  
+  see http://www.nec-labs.com/uploads/images/Department-Images/MediaAnalytics/papers/nips16_npairmetriclearning.pdf
+
+  Npair loss requires paired data. Npair loss has two parts, the first part is L2
```
Comment (haowang101779990) [inline python/paddle/fluid/layers/nn.py]:
```suggestion
  Npair loss requires paired data. Npair loss has two parts: the first part is L2 
```

File: python/paddle/fluid/layers/nn.py [hunk 5]
```diff
@@ -10560,3 +10561,52 @@ def tree_conv(nodes_vector,
...
+
+
+def npair_loss(anchor, positive, labels, l2_reg=0.002):
+    '''
+  **Npair Loss Layer**
+  
+  see http://www.nec-labs.com/uploads/images/Department-Images/MediaAnalytics/papers/nips16_npairmetriclearning.pdf
+
+  Npair loss requires paired data. Npair loss has two parts, the first part is L2 
+  regularizer on the embedding vector, the second part is cross entropy loss which
```
Comment (haowang101779990) [inline python/paddle/fluid/layers/nn.py]:
```suggestion
  regularizer on the embedding vector; the second part is cross entropy loss which
```

Please use semicolon ';' to introduce equal level of parts.

**Review by haowang101779990 — COMMENTED** [review]
Thanks for your preview of API. I have reviewed and there are some minors to fix. :) 👍 @shanyi15

**Review by wanghaoshuang — DISMISSED** [review]
LGTM.

**Comment by shanyi15** [comment]
please preview again to make sure the link is corrected. refer to
https://github.com/PaddlePaddle/Paddle/pull/15949#discussion_r261040652

**Review by shanyi15 — DISMISSED** [review]
format and content of comment LGTM

**Review by shanyi15 — DISMISSED** [review]

**Review by panyx0718 — DISMISSED** [review]

**Comment by shanyi15** [comment]
请问example的内容可以丰富一些吗？

File: python/paddle/fluid/layers/nn.py [hunk 6]
```diff
@@ -10652,3 +10652,52 @@ def tree_conv(nodes_vector,
...
+    labels(Variable): 1-D tensor. shape=[batch_size]
+    l2_reg(float32): L2 regularization term on embedding vector, default: 0.002
+
+  Returns:
+    npair loss(Variable): return npair loss, shape=[1]
+
+  Examples:
+    .. code-block:: python
+
+       npair_loss = fluid.layers.npair_loss(anchor, positive, labels, l2_reg)
```
Comment (haowang101779990) [inline python/paddle/fluid/layers/nn.py]:
```suggestion
       npair_loss = fluid.layers.npair_loss(anchor, positive, labels, l2_reg)
```
可否在示例里面添加关于定义anchor, positive, labels, l2_reg的语句并加以简单注释，这样示例
才能真正起到效果。

Comment (ceci3) [reply python/paddle/fluid/layers/nn.py]: 好的

**Review by panyx0718 — DISMISSED** [review]

**Review by shanyi15 — DISMISSED** [review]
LGTM

File: python/paddle/fluid/layers/nn.py [hunk 7]
```diff
@@ -10652,3 +10652,59 @@ def tree_conv(nodes_vector,
...
+       npair_loss = fluid.layers.npair_loss(anchor, positive, labels, l2_reg = 0.002)
+  '''
+    Beta = 0.25
+    batch_size = labels.shape[0]
+
+    labels = reshape(labels, shape=[batch_size, 1], inplace=True)
+    labels = expand(labels, expand_times=[1, batch_size])
+
+    from .control_flow import equal
+    from .ops import square
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py]: don't import here.

Comment (ceci3) [reply python/paddle/fluid/layers/nn.py]: done

**Review by panyx0718 — DISMISSED** [review]

**Review by shanyi15 — DISMISSED** [review]

**Comment by ceci3** [comment]
文档预览如下：
屏幕快照 2019-03-08 下午1 48 44

**Review by shanyi15 — APPROVED** [review]
LG comment

File: python/paddle/fluid/layers/nn.py [hunk 8]
```diff
@@ -10652,3 +10652,60 @@ def tree_conv(nodes_vector,
     else:
         pre_activation = out
     return helper.append_activation(pre_activation)
+
+
+from .ops import square
+from .control_flow import equal
```
Comment (panyx0718) [inline python/paddle/fluid/layers/nn.py:10658]: this usually mean the follow layer doesn't belong to this file or some refactoring is needed.

**Review by panyx0718 — APPROVED** [review]
LG API
